autoload/go/lsp.vim

```diff
@@ -0,0 +1,302 @@
+    call go#util#EchoError('This feature requires either Vim 8.0.0087 or newer with +job or Neovim.')
+    return
+  endif
+
+  " job is the job used to talk to the backing instance of gopls.
+  " ready is 0 until the initialize response has been received. 1 afterwards.
+  " queue is messages to send after initialization
+  " last_request_id is id of the most recently sent request.
+  " buf is unprocessed/incomplete responses
+  " handlers is request id -> callback
+  let l:lsp = {
+        \ 'job':  '',
+        \ 'ready': 0,
+        \ 'queue': [],
+        \ 'last_request_id': 0,
+        \ 'buf': '',
+        \ 'handlers': {},
+        \ }
+
+  function! l:lsp.readMessage(data) dict abort
+    let l:responses = []
+    let l:rest = a:data
+
+    while 1
+      " Look for the end of the HTTP headers
+      let l:body_start_idx = matchend(l:rest, "\r\n\r\n")
+
+      if l:body_start_idx < 0
+        " incomplete header
+        break
+      endif
+
+      " Parse the Content-Length header.
+      let l:header = l:rest[:l:body_start_idx - 4]
+      let l:length_match = matchlist(
+      \   l:header,
+      \   '\vContent-Length: *(\d+)'
+      \)
+
+      if empty(l:length_match)
+        " TODO(bc): shutdown gopls?
+        throw "invalid JSON-RPC header:\n" . l:header
+      endif
+
+      " get the start of the rest
+      let l:rest_start_idx = l:body_start_idx + str2nr(l:length_match[1])
+
+      if len(l:rest) < l:rest_start_idx
+        " incomplete response body
+        break
+      endif
+
+      if go#util#HasDebug('lsp')
+        let g:go_lsp_log = add(go#config#LspLog(), {
+              \ 'response': l:rest[:l:rest_start_idx - 1]
+        \ })
+      endif
+
+      let l:body = l:rest[l:body_start_idx : l:rest_start_idx - 1]
+      let l:rest = l:rest[l:rest_start_idx :]
+
+      try
+        " add the json body to the list.
+        call add(l:responses, json_decode(l:body))
+
+        " TODO(bc): set status line stuff similarly to how job.vim handles it.
+      catch
+        " TODO(bc): log the message and/or show an error message.
+      finally
+        " intentionally left blank.
+      endtry
+    endwhile
+
+    return [l:rest, l:responses]
+  endfunction
+
+  function! l:lsp.handleMessage(ch, data) dict abort
+      let self.buf .= a:data
+
+      let [self.buf, l:responses] = self.readMessage(self.buf)
+
+      if self.ready is 0
+        for l:response in l:responses
+          " TODO(bc): handle notifications that aren't InitializeResult (e.g.
+          " window/showMessage).
+          call self.handleInitializeResult(l:response)
+        endfor
+      endif
+
+      if self.ready is 1
+        for l:response in l:responses
+          if has_key(l:response, 'id') && has_key(self.handlers, l:response.id)
+            try
+              if has_key(l:response, 'error')
+                call go#util#EchoError(l:response.message)
```

Comment:
@bhcleek Should this be `call go#util#EchoError(l:response.error)`? I'm getting this when the response has `error`:
```
E716: Key not present in Dictionary: message)
E116: Invalid arguments for function go#util#EchoError
```

btw, thank you for adding the gopls support!

Reply:
yes! I've updated master accordingly.

Reply:
Thank you!!